### PR TITLE
Widget::size-allocate

### PR DESCRIPTION
**Pull request by EPashkin**
Part of https://github.com/gtk-rs/gir/issues/244
cc @gkoz 
Alternative for manual `connect_size_allocate` will be move rust proxy of `Rectangle` to `cairo`.

WIP

### CODE REVIEW

**Comment by EPashkin** [comment]
Maybe `gtk_tool_palette_get_drag_item` has not optional return.
Other that that ready for merge

File: src/widget.rs [hunk 1]
```diff
@@ -36,6 +37,8 @@ pub trait WidgetExtManual {
 
     fn connect_map_event<F: Fn(&Self, &Event) -> Inhibit + 'static>(&self, f: F) -> u64;
 
+    fn connect_size_allocate<F: Fn(&Self, &RectangleInt) + 'static>(&self, f: F) -> u64;
```
Comment (gkoz) [inline src/widget.rs]: It's already possible to refer to the rectangle type as `Allocation` and the generated widget code does.

File: src/widget.rs [hunk 2]
```diff
@@ -99,3 +110,10 @@ where T: IsA<Widget> {
     callback_guard!();
     f(&Widget::from_glib_none(this).downcast_unchecked(), &from_glib_none(event)).to_glib()
 }
+
+unsafe extern "C" fn rectangle_trampoline<T>(this: *mut ffi::GtkWidget, allocation: *mut RectangleInt,
```
Comment (gkoz) [inline src/widget.rs]: It'd be cleaner to use the ffi type (`cairo_rectangle_int_t`) here.

**Comment by EPashkin** [comment]
Found source problem with generating `size-allocate`: generated trampoline has allocation without reference,  I try to fix and enable generation in this PR.
```
unsafe extern "C" fn size_allocate_trampoline<T>(this: *mut ffi::GtkWidget, allocation: ffi::GtkAllocation, f: glib_ffi::gpointer)
```

**Comment by EPashkin** [comment]
Moved unrelated part to https://github.com/gtk-rs/gtk/pull/341

**Comment by EPashkin** [comment]
..due fixing gir take some time

**Comment by EPashkin** [comment]
Strange even after fix
```
unsafe extern "C" fn size_allocate_trampoline<T>(this: *mut ffi::GtkWidget, allocation: *mut ffi::GtkAllocation, f: glib_ffi::gpointer)
where T: IsA<Widget> {
    callback_guard!();
    let f: &Box_<Fn(&T, &Allocation) + 'static> = transmute(f);
    f(&Widget::from_glib_none(this).downcast_unchecked(), &from_glib_none(allocation))
}
```
I get errors
```
D:/eap/rust/0/gtk\src\auto\widget.rs:2815:60: 2815:74 error: the trait bound `rectangle::Rectangle: glib::translate::FromGlibPtr<*mut cairo_ffi::cairo_rectangle_int_t>` is not satisfied [E0277]
D:/eap/rust/0/gtk\src\auto\widget.rs:2815     f(&Widget::from_glib_none(this).downcast_unchecked(), &from_glib_none(allocation))
                                                                                                     ^~~~~~~~~~~~~~
D:/eap/rust/0/gtk\src\auto\widget.rs:2815:60: 2815:74 help: run `rustc --explain E0277` to see a detailed explanation
D:/eap/rust/0/gtk\src\auto\widget.rs:2815:60: 2815:74 note: required by `glib::translate::from_glib_none`
```
Manual trampoline currently use transmute, not `from_glib_xxx`
```
-    f(&Widget::from_glib_none(this).downcast_unchecked(), transmute(allocation));
+    let f: &Box_<Fn(&T, &Allocation) + 'static> = transmute(f);
+    f(&Widget::from_glib_none(this).downcast_unchecked(), &from_glib_none(allocation))
```

**Comment by gkoz** [comment]
Oh right, the `FromGlibPtr` implementation is not there, we didn't need it previously.

**Comment by EPashkin** [comment]
Seems it time to make choice:
- using `cairo`s `RectangleInt` anywhere
- update `gtk` `Rectangle` with `FromGlib`
- move `gtk` `Rectangle` to `gdk` with adding `FromGlib` (as you say that in next version it has separeted from `cairo`s)

**Comment by gkoz** [comment]
Now that cairo has its `RectangleInt` switched to conversions, it should be possible to reexport it as `gdk::Rectangle` and `gtk::Allocation` and use the respective names in either crate.

**Comment by gkoz** [comment]
We might have an issue when after updating gir-files `GdkRectangle` in the sys crates stops being an alias to `cairo_rectangle_int_t`...

**Comment by gkoz** [comment]
So I suppose it's safer to make `gdk::Rectangle` an independent type and simply implement `From`/`Into` with `cairo::RectangleInt`.

**Comment by EPashkin** [comment]
Updated with temporary regen

**Comment by EPashkin** [comment]
Interesting, `cairo::RectangleInt` has all fields private. While build error will be fixed after  https://github.com/gtk-rs/gdk/pull/119, IMHO better make field public.

**Comment by gkoz** [comment]
Uh, what a silly mistake. :blush:

**Comment by EPashkin** [comment]
Ready for merge

**Comment by GuillaumeGomez** [comment]
:+1:

**Comment by EPashkin** [comment]
It builds even without `gdk` part. Its good.

**Comment by gkoz** [comment]
Thanks!

**Pull request merged**